File: examples/custom_op/inner_product/gradtest.jl
using ADCME
using PyCall
if Sys.isapple()
    py"""
    import tensorflow as tf
    from tensorflow.python.framework import ops
    inner_product_module = tf.load_op_library('build/libinner_product.dylib')
    inner_product_grad_module = tf.load_op_library('build/libinner_product_grad.dylib')
    @ops.RegisterGradient("InnerProduct")
    def _inner_product_grad_cc(op, grad):
        return inner_product_grad_module.inner_product_grad(grad, *op.inputs)
    """
else
    py"""
    import tensorflow as tf
    from tensorflow.python.framework import ops
    inner_product_module = tf.load_op_library('build/libinner_product.so')
    inner_product_grad_module = tf.load_op_library('build/libinner_product_grad.so')
    @ops.RegisterGradient("InnerProduct")
    def _inner_product_grad_cc(op, grad):
        return inner_product_grad_module.inner_product_grad(grad, *op.inputs)
    """
end

inner_product = py"inner_product_module.inner_product"

x = constant(rand(100,1), dtype=Float32)
W = constant(rand(100,100), dtype=Float32)
g = inner_product(x, W)
dg = gradients(sum(W*x), x) 
sess = Session()
run(sess, [g, dg])

File: examples/custom_op/inner_product/gradtest.py
#!/usr/bin/env python3
"""
Tests for the inner product Tensorflow operation.

.. moduleauthor:: David Stutz
"""

import unittest
import numpy as np
import tensorflow as tf
from tensorflow.python.framework import ops
# from tensorflow.python.ops import array_ops
# from tensorflow.python.ops import sparse_ops
inner_product_module = tf.load_op_library('build/libinner_product.dylib')
inner_product_grad_module = tf.load_op_library('build/libinner_product_grad.dylib')
@ops.RegisterGradient("InnerProduct")
def _inner_product_grad_cc(op, grad):
    return inner_product_grad_module.inner_product_grad(grad, op.inputs[0], op.inputs[1])


sess = tf.Session()
res = inner_product_module.inner_product([[1], [2]], [[1, 2], [3, 4]])
print(sess.run(res))

x = tf.constant(np.asarray([[1.], [2.]]).astype(np.float32))
W = tf.constant(np.asarray([[1, 2], [3, 4]]).astype(np.float32))
y = inner_product_module.inner_product(x, W)
grad_y = tf.gradients(y[1], W)[0]
sess.run(tf.global_variables_initializer())
print(sess.run(grad_y))
sess.close()
